### PR TITLE
Issue 716: Fix error in calculation of grid average inside density_from_Universe

### PR DESCRIPTION
**Pull request by Balasubra**
This pull request is related to issue #716

### CODE REVIEW

**Comment by jbarnoud** [comment]
Somehow, this breaks a test but I do not see why:

```
======================================================================
FAIL: test_density_from_Universe_sliced (MDAnalysisTests.analysis.test_density.Test_density_from_Universe)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/pyenv/lib/python2.7/site-packages/MDAnalysisTests/analysis/test_density.py", line 135, in test_density_from_Universe_sliced
    start=1, stop=-1, step=2,
  File "/home/travis/miniconda/envs/pyenv/lib/python2.7/site-packages/MDAnalysisTests/analysis/test_density.py", line 117, in check_density_from_Universe
    err_msg="mean density does not match")
  File "/home/travis/miniconda/envs/pyenv/lib/python2.7/site-packages/numpy/testing/utils.py", line 513, in assert_almost_equal
    raise AssertionError(_build_err_msg())
AssertionError: 
Arrays are not almost equal to 7 decimals mean density does not match
 ACTUAL: 0.016764270747693617
 DESIRED: 0.006705708879402314
```

@orbeckst it seems you wrote the test.; any idea?

**Comment by richardjgowers** [comment]
I think the old values are wrong. The new values match the unsliced result, which makes sense because the density shouldn't change that much depending on number of frames

File: package/MDAnalysis/analysis/density.py [hunk 1]
```diff
@@ -514,7 +514,9 @@ def current_coordinates():
 
         h[:], edges[:] = np.histogramdd(coord, bins=bins, range=arange, normed=False)
         grid += h  # accumulate average histogram
-    n_frames = u.trajectory.n_frames
+   
+    start, stop, step = u.trajectory.check_slice_indices(start, stop, step)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/density.py]: The slice check should come before the loop over frames – we (I?) forgot it.

**Comment by orbeckst** [comment]
@richardjgowers is most likely correct and the old value was wrong.
Change the test fixture to the new values.

**Comment by hainm** [comment]
may be add more description to the title. 

"Balasubra issue #716" seems to be not really informative in the future.

**Comment by orbeckst** [comment]
@Balasubra you will also need to change the density test case.

**Comment by Balasubra** [comment]
@orbeckst  what do i change in density test case, i dnt get it.

**Comment by orbeckst** [comment]
There's a test case in `analysis/test_density.py`  failing (look at the output from Travis https://travis-ci.org/MDAnalysis/mdanalysis/builds/108851425 ):

```
======================================================================
FAIL: test_density_from_Universe_sliced (MDAnalysisTests.analysis.test_density.Test_density_from_Universe)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/pyenv/lib/python2.7/site-packages/MDAnalysisTests/analysis/test_density.py", line 135, in test_density_from_Universe_sliced
    start=1, stop=-1, step=2,
  File "/home/travis/miniconda/envs/pyenv/lib/python2.7/site-packages/MDAnalysisTests/analysis/test_density.py", line 117, in check_density_from_Universe
    err_msg="mean density does not match")
  File "/home/travis/miniconda/envs/pyenv/lib/python2.7/site-packages/numpy/testing/utils.py", line 513, in assert_almost_equal
    raise AssertionError(_build_err_msg())

AssertionError: 
Arrays are not almost equal to 7 decimals mean density does not match
 ACTUAL: 0.016764270747693617
 DESIRED: 0.006705708879402314
```

In this test, replace the existing number **0.006705708879402314** with the number that was computed during the test run, **0.016764270747693617**. We want to do this because the old number was incorrect and the new one looks like the correct one.

**Comment by orbeckst** [comment]
Also add an entry to CHANGELOG, referencing #716 and add your name to the 0.14 CHANGELOG entry (if it's not already there).

Please squash the three commits into one (have a look at `git rebase -i` or ask).

**Comment by Balasubra** [comment]
@orbeckst  Please bear with me, I am getting used to github in a slow phase.
I have updated CHANGELO, test_density.py  in branch  Balasubra-patch-1. Now i have five commits under the pull request pull request #716

Before that i made a mess that i made changes to CHANGELOG, test_density.py in a new path called Balasubra-patch-2.

I am trying to figure out a) to delete this new patch-2 and b) also to first three commits in density.py into one as you suggested.

**Comment by jbarnoud** [comment]
@Balasubra 
You can delete a branch with `git branch -D name-of-the-branch`.

You can read how to squash commits in the [git doc](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#Squashing-Commits). You want to squash all your commits to your first one.

File: package/MDAnalysis/analysis/density.py [hunk 2]
```diff
@@ -501,6 +501,7 @@ def current_coordinates():
     pm = ProgressMeter(u.trajectory.n_frames, interval=interval, quiet=quiet,
                        format="Histogramming %(n_atoms)6d atoms in frame "
                        "%(step)5d/%(numsteps)d  [%(percentage)5.1f%%]\r")
+    start, stop, step = u.trajectory.check_slice_indices(start, stop, step)
```
Comment (Balasubra) [inline package/MDAnalysis/analysis/density.py]: I am now wondering, why start and stop should be None by default. May be start=0 and stop=len(u.trajectory).
Suppose the user simply want to process every 10th frame, he may just give the step value. If not every time one should remember to do give start,stop values even if the use simply needs skipping (i.e. step).

Comment (richardjgowers) [inline package/MDAnalysis/analysis/density.py]: Yeah using None will default them to those values.  The slice `a[::10]` is identical to `a[None:None:10]` in Python

Comment (jbarnoud) [inline package/MDAnalysis/analysis/density.py]: This reproduce how slicing works everywhere in Python. If start, end, or step is `None`, then the correct value will be deduced based on the rest. Most of the time, start will be automatically set to 0, end to `len(u.trajectory)`, and step to 1. But if you want to read your whole trajectory backward, then you will set step to -1, you will left start and end to `None`, and their correct value will be deduced.

Look at how `range` works. It is the same here.

**Comment by Balasubra** [comment]
@orbeckst  sorry that i cudnt figure out to merge commits to one.
Could any of you, based on your experience, suggest any good gui client for using github in linux. I installed smartgit and tried to play with commits but i cudnt get to do it correctly.

**Comment by kain88-de** [comment]
@Balasubra [this is a good tutorial](http://rypress.com/tutorials/git/rebasing) on rebasing.  Go to the part telling about cleaning up git-history. It gives you a more indepth explanation how the git UI for interactive rebases work. Every UI to git will prompt you with a similar interface.

**Comment by jbarnoud** [comment]
@Balasubra Don't worry about squashing the commits. It is not a trivial operation, I will take care of it.

**Comment by jbarnoud** [comment]
I squashed the commits and merged them manually. So this is merged, even if it is not displayed as such.

@Balasubra Thanks for fixing the bug.

**Comment by jbarnoud** [comment]
(See commit 78992cffd40682ffa644c486ccbc638232497ae4)

**Comment by kain88-de** [comment]
@jbarnoud github should recognize the merge when you force the creation of a merge commit. Then there is no need to close it manual

**Comment by jbarnoud** [comment]
I thought so but it did not work. Probably because of the rebase. Also I 
changed the commit message, that probably did not help.

**Pull request closed without merging**